### ISSUE TITLE
[low-bit optim] Add COAT optimizer

### ISSUE THREAD

**Issue by gau-nernst** [pytorch/ao#1190]
Paper: https://arxiv.org/abs/2410.19313
Code: https://github.com/NVlabs/COAT (not available yet)

Seems like we already have most of the building blocks. The only new logic is "dynamic range expansion"

![](https://github.com/user-attachments/assets/d3c4da8d-82e7-486e-81d9-7b96fe842b66)

**Comment by MirMustafaAli** [comment]
I would like to work on this @gau-nernst!!

**Comment by gau-nernst** [comment]
@MirMustafaAli Go ahead and submit a PR 😄. Let me know if you face any problems.

**Comment by MirMustafaAli** [comment]
@gau-nernst Which section should i look for to implement "dynamic range expansion"?. According to my understanding of repo it must be in float8 folder as it's aimed at hopper architecture utilizing type float8.  _Any pointers, PR's and reference methods which i can follow would be very much helpful for me_.

**Comment by gau-nernst** [comment]
You can park it under `torchao/prototype/low_bit_optim`. The `float8/` folder is more for training stuff (fp8 matmul).

You can extend our current `OptimStateFp8`. See https://github.com/pytorch/ao/blob/000a49026459dd1dadf5ca34322d98e7b1680250/torchao/prototype/low_bit_optim/subclass_fp8.py. I think we only need to change the `quantize_fp8` function.

Another option is to create a separate optimizer. See https://github.com/pytorch/ao/blob/000a49026459dd1dadf5ca34322d98e7b1680250/torchao/prototype/low_bit_optim/adam.py. You can wrap all of the logic in a functional way (see `single_param_adam()` in the link above), and add boilerplate code for the optimizer (i.e. init optim states, call the functional optim step with torch.compile...)

**Comment by MirMustafaAli** [comment]
Thanks!! Will work on your advice.

**Comment by shreynath** [comment]
Hi! I'm Shreyaan, a Computer Engineering Student at Texas A&M getting started with open source contributions and I'd like to take a stab at this. From the thread it sounds like the cleanest path is extending OptimStateFp8 in torchao/prototype/low_bit_optim/subclass_fp8.py to implement the dynamic range expansion step from the COAT paper. Is this still available, and is the subclass_fp8.py extension approach still preferred over writing a standalone optimizer? The NVLABS code repo being unavailable means I'd be working from the paper directly — is that sufficient?

**Comment by ATMmonitor667** [comment]
Hi @gau-nernst, I'd like to pick this up if it's available. 
torchao already has an FP8 optimizer, and COAT extends that with a dynamic range expansion step that lowers the accuracy loss when the optimizer states are compressed. My plan is to add an FP8 optimizer state variant that includes that step, plus CoatAdam and CoatAdamW classes matching the existing optimizer API. Since the Adam step runs in fp32 and only interacts with the state through lerp and copy, the core update logic shouldn't need to change.
One question before I start: do you prefer new optimizer classes, or an option on the existing AdamFp8 to enable this? I'll follow the NVlabs COAT reference for the exact formula and keep the scope to optimizer states only.
